packages/reason-css-parser/ppx/EmitType.re

```diff
@@ -0,0 +1,301 @@
+        | Property_type(name) =>
+          let name = property_value_name(name) |> value_name_of_css;
+          let params = [ptyp_constr(txt @@ Lident(name), [])];
+          (first_uppercase(name), false, params);
+        | _ => raise(Unsupported)
+        };
+      apply_modifier(multiplier, type_, is_constructor, params);
+    };
+
+    let combinator_op = (kind, values) => {
+      switch (kind) {
+      | Xor =>
+        List.map(
+          fun
+          | Terminal(kind, multiplier) => terminal_op(kind, multiplier)
+          | _ => raise(Unsupported),
+          values,
+        )
+      | _ => raise(Unsupported)
+      };
+    };
+
+    switch (value) {
+    | Terminal(kind, multiplier) =>
+      mk_typ(type_name) @@ [terminal_op(kind, multiplier)]
+    | Combinator(kind, values) =>
+      mk_typ(type_name) @@ combinator_op(kind, values)
+    // | Group(value, multiplier) => group_op(value, multiplier)
+    // | Function_call(name, value) => function_call(name, value)
+    | _ => abstract_type(type_name)
+    };
+  };
+};
+
+let extract_ppx_content = (exp: Parsetree.expression) => {
+  switch (exp.pexp_desc) {
+  | Pexp_extension((
+      _,
+      PStr([
+        {
+          pstr_desc:
+            Pstr_eval(
+              {
+                pexp_desc: Pexp_constant(Pconst_string(value, loc, _delim)),
+                _,
+              },
+              _attrs,
+            ),
+          _,
+        },
+      ]),
+    )) => (
+      value,
+      loc,
+    )
+  | _ => failwith("expected a ppx extension")
+  };
+};
+
+
+let standard_types = {
+
+  open Ppxlib.Ast_builder.Default;
+
+  let abstract_type = name => {
+      type_declaration(
+        ~loc=Location.none,
+        ~name={txt: name, loc: Location.none},
+        ~params=[],
+        ~cstrs=[],
+        ~private_=Public,
+        ~manifest=None,
+        ~kind=Ptype_abstract,
+      );
+  };
+
+  [
+    abstract_type("integer"),
+    abstract_type("number"),
+    abstract_type("length"),
+    abstract_type("angle"),
+    abstract_type("time"),
+    abstract_type("frequency"),
+    abstract_type("resolution"),
+    abstract_type("percentage"),
+    abstract_type("ident"),
+    abstract_type("custom_ident"),
+    abstract_type("any_value"),
+    // abstract_type("string"), already represented by OCaml string type
+    abstract_type("url"),
+    abstract_type("hex_color"),
+    abstract_type("interpolation"),
+    abstract_type("flex_value"),
+
+  // Not at Standard.re but required by genereted code, should they live here?
+    abstract_type("hash_token"),
+    abstract_type("dimension"),
+    abstract_type("an_plus_b"),
+  ]
+}
+
+let extract_variable_name = (pat: Parsetree.pattern) => {
+  switch (pat.ppat_desc) {
+  | Ppat_var({txt, _}) => txt
+  | _ => failwith("expected variable name")
+  };
+};
+
+let gen_type = (binding: Parsetree.value_binding) => {
+  let name = extract_variable_name(binding.pvb_pat);
+  let (payload, loc) = extract_ppx_content(binding.pvb_expr);
```

Comment:
These 2 functions can raise and there's seems to no be any reason to not have the switch expressions together inlined here. Thoguhts?

Reply:
I think that the inlined version is not very expressive.
```reason
  let (name, payload, loc) =
    switch (binding) {
    | {
        pvb_pat: {ppat_desc: Ppat_var({txt, _}), _},
        pvb_expr: {
          pexp_desc:
            Pexp_extension((
              _,
              PStr([
                {
                  pstr_desc:
                    Pstr_eval(
                      {
                        pexp_desc:
                          Pexp_constant(Pconst_string(value, loc, _delim)),
                        _,
                      },
                      _attrs,
                    ),
                  _,
                },
              ]),
            )),
          _,
        },
        _,
      } => (
        txt,
        value,
        loc,
      )
    | _ => failwith("Error when extracting CSS spec content")
    }
```

Reply:
It's very common to have this kind of nested statement on ppxes.